Multi-entity save UI: Add Discard Changes panel

ockham opened the pull request:
## Description

**This needs some UX feedback from designers, see https://github.com/WordPress/gutenberg/pull/36185#issuecomment-961466024**

WIP, needs more work.

Add a second step to the multi-entity save UI which offers the user to discard any changes they made and didn't opt to save. This implements the second part of https://github.com/WordPress/gutenberg/issues/31456#issuecomment-888218396 (and thus closes #31456), after https://github.com/WordPress/gutenberg/pull/35933 implemented the first part.

## How has this been tested?

- Use the TT1 Blocks theme.
- Go to the Site Editor, and select the Single Post Template to edit.
- Make a few minor changes: Change the site title and tagline; insert a paragraph block with some text near the post content; change the address in the site footer.
- Click 'Save'.
- Select a few (but not all!) changes to save, and click 'Save' again.
- You're presented with the "discard changes" panel.
- _To be continued_

## Screenshots

Current state per this PR:

![image](https://user-images.githubusercontent.com/96308/140178757-4f61d621-75bc-4c6a-b8c2-60812b38fd3f.png)

## Types of changes
New feature

## TODO

- [x] Figure out what to do with undo stack, see https://github.com/WordPress/gutenberg/pull/36185#issuecomment-961466024
- [x] Correctly wire buttons, checkboxes, etc
- [x] Split into separate files for save and discard UI
- [x] Add unit tests for newly added actions, reducers
- [x] Rebase, carry over changes from #36212 to `discardEditedEntryRecords` if needed.
- [x] Fix isSaving filtering for site entity record edits (title, tagline, etc -- might need `title` instead of `key`?)
- [x] Add loading state (pulsate entities)?
- [x] Add snackbar after discard (see https://github.com/WordPress/gutenberg/pull/36185#issuecomment-967239470)
- [ ] Update e2 test.

ockham committed:
Multi-entity save UI: Add Discard Changes panel

paaljoachim commented:
Looks good Bernie! Thank you!

The process is like so:
User clicks the Save button sees options of various areas to save. Decides to uncheck two areas.

![Screenshot 2021-11-04 at 11 58 16](https://user-images.githubusercontent.com/5323259/140302363-20dd96b2-3457-4725-9c8e-bf6e0c8464d6.jpg)

Clicks Save again.
Clicks Front Page and Another post to discard the changes.

![Screenshot 2021-11-04 at 11 59 21](https://user-images.githubusercontent.com/5323259/140302512-d6fb05a3-fbb0-4f4b-9d9b-e3d0630cf4bb.jpg)

(I am also seeing a "Continue editing" button further down.)

Clicks the "Discard changes" button. Panel is closed and user is returned to working on the layout.
The Save button is still active and not greyed out. 

The question comes up.... what does discard actually do?
I assume that discard is the same as revert the current selected area back to what it was the last time it was saved. 

@javierarce @critterverse

ockham commented:
> Clicks the "Discard changes" button. Panel is closed and user is returned to working on the layout.
> The Save button is still active and not greyed out.
>
> The question comes up.... what does discard actually do?
> I assume that discard is the same as revert the current selected area back to what it was the last time it was saved.

Yeah, that's what I also think it means. I just haven't implemented that behavior yet 😅 (that's the "Correctly wire buttons, checkboxes, etc" item in my little todo list in the PR desc 😉)

ockham committed:
Discard changes i1

ockham committed:
Make work for site entities

ockham commented:
Took me a while to figure out how to implement discarding changes; turns out we didn't have any actions to reset an entity record's edits (that's what discarding changes translates to in Gutenberg parlance), so I had to implement something.

This still needs a bit of polish plus some test coverage etc, which I could probably finish tomorrow 🤞 

---

The most important question however is what to do with the undo stack: Typically, every change (edit) creates an entry on the undo stack, which allows the user to go back in time step by step.

AFAICT, the feature we're introducing here -- discarding changes on a per-entity basis -- doesn't seem to be entirely compatible with that pattern:

- It doesn't just affect one change, but _all_ changes to a given thing ("entity record") since it was last saved.
- Those changes might not have happened subsequently, but interspersed with changes to other entity records.

Here's an example: After freshly opening some template in the site editor, the user does the following the site editor, in order:

- Changes some text in the header.
- Changes the site title.
- Adds an image block to the header.
- Changes the site tagline.
- Changes some text in the footer.
- Adds some more text to the header.
- Changes the site title once again.

Using undo, we could obviously go back step by step.

Now let's say the user clicks Save, but decides to only save the text change to the footer template part. This means that our new discard panel then asks them if they want to discard the other changes that they made to the header, site title, and tagline. Let's say the user chooses to discard the changes to the header and the site tagline: Those changes are then rolled back, whereas the ones to the site title are kept.

Now what should happen to the undo stack (and button)? What would be most intuitive for the user? I can think of some options.
1. Discarding the changes is recorded on the undo stack itself, so if the user clicks undo, the changes that they just dismissed are brought back.
   - This probably works if the user only discarded changes to _one_ thing, but it might not be possible if they discarded changes to multiple things (like in our example). (After all, it's also not possible in manual editing to, say, modify the header and site tagline _at once_ -- and that fact is somewhat reflected by the undo stack implementation.)
2. Disable the undo button (i.e. clear the undo stack). Might be conceptually clean, but seems like a bit of a sledgehammer approach to me.
3. Remove the discarded changes from the undo stack, and keep the other ones. In our example, this would mean that after discarding the changes to header and site tagline, the undo stack would shrink to

- Changes the site title.
- Changes some text in the footer.
- Changes the site title once again.

Would that feel intuitive to the user, if they started clicking the undo button? Are there any more intuitive options that I'm missing?

@javierarce @critterverse

ockham commented:
Forgot to say, there's also the concept of _transient_ edits (which don't create entries on the undo stack, nor do they dirty an entity's state), but I believe that they're "merged" into the next item on the undo stack, and I don't think we want that(?) (Though I have to say my understanding of transient edits might be a bit fuzzy.)

ockham committed:
Remove commented-out code

talldan commented:
@ockham I face a similar conundrum over on https://github.com/WordPress/gutenberg/pull/36030 (also some related discussion on https://github.com/WordPress/gutenberg/pull/36027). 

For removing an entity, I was going for the third option you outline.

For a discard though, it does seem like something a user might be able to undo.

ockham committed:
No need to set an actual value

ockham committed:
Change implementation to use EDIT_ENTITY_RECORD

ockham committed:
Add unit test

ockham commented:
> @ockham I face a similar conundrum over on #36030 (also some related discussion on #36027).
> 
> For removing an entity, I was going for the third option you outline.
> 
> For a discard though, it does seem like something a user might be able to undo.

Thanks a lot @talldan, I was unaware of #36030 -- it's indeed a similar problem.

I've been giving this some more thought. We might not necessarily need to allow the user to undo the discard -- it's after all highlighted as a destructive option.

For now, I've decided to try to implement the discard through `EDIT_ENTITY_RECORD` (with `meta: { undo: undefined }`). I'll see how far this can get us...

ockham committed:
Change implementation to use EDIT_ENTITY_RECORD

ockham committed:
Fix unit test

ockham committed:
Add unit test for __experimentalResetEditedEntityRecord

ockham committed:
Close panel upon clicking either button

ockham committed:
Move DiscardEntityChangesPanel to dedicated file

ockham committed:
Fix some criteria

ockham committed:
Simplify

ockham committed:
Some polish

ockham committed:
Style fix

ockham commented:
I think this is coming together. I need a bit of design input for one transition, though: After clicking 'Save', GB takes a moment to actually save the changed entities. Currently, it already moves on to the 'Discard' screen, and for a brief moment, all entity change checkboxes are carried over (even the ones that were just saved):

![multi-entity save discard](https://user-images.githubusercontent.com/96308/141195115-b2b870dc-e777-4921-931f-c15ce4068a1c.gif)

As a minimum, we should probably disable the interactive elements (checkboxes, "Discard changes", and "Continue Editing" buttons) in the sidebar while saving.

I'd appreciate some guidance/mockups from a designer! 🙏

paaljoachim commented:
Testing the PR with a "ops I made a change to a Reusable block" situation. As this part of the feature will be extremely helpful for all those folks making accidental changes to a Reusable block.

This is what I see:

https://user-images.githubusercontent.com/5323259/141370917-2c56f992-9f5c-4edc-b1e8-968f83adf1a3.mp4

Walking through the scenario.
1- Make a change to a Reusable block.
2- Notice the dot show up beside the active Update... button.
3- Click Update.
4- Save panel is seen. 
5- Unclick the checkbox to not save the Reusable block change. 
6- Click Save.
7- Notice that for a moment both checkboxes are carried over (also seen in Bernies video above).
I would suggest to have a spinner animation https://wordpress.github.io/gutenberg/?path=/story/components-spinner--default to show that something is happening instead of showing all the checkboxes. Using it for a moment before moving on to only show the discarded checkboxes.

8- Click the checkbox and then the Discard changes button to discard the option.
9- Exits the Update/Save panel procedure.

javierarce commented:
> The question comes up.... what does discard actually do?
> I assume that discard is the same as revert the current selected area back to what it was the last time it was saved.

Sorry for my late reply, @paaljoachim & @ockham… yes, that's exactly what the button does.

Regarding the rest of the PR, it looks good! I only miss the success message indicating that the previous save action was successful. Otherwise, the "What's next" title will be confusing. That title is also important because we aren't showing the snack bar.

I also think that we could remove the "Continue editing" button. It's too far away from the other actions (the discard button and the close button at the top) and competes with them.

If you all agree to make this change, we should then update the message to something like:

> **What's next?**
> Your template still has some unsaved changes.
> You can select them and discard their changes now, or close the panel and deal with them later.

javierarce commented:
> I think this is coming together. I need a bit of design input for one transition, though: After clicking 'Save', GB takes a moment to actually save the changed entities. Currently, it already moves on to the 'Discard' screen, and for a brief moment, all entity change checkboxes are carried over (even the ones that were just saved)

As @paaljoachim suggests, could we do the opposite? Hide every item and just show the ones that make sense to show?

javierarce commented:
Something important we should also include is a snackbar message after the user discards the changes.

Some suggestions for the message (depending on the number of items reverted):

- "All changes were discarded"
- "Some changes were discarded"

Another idea. In case there's just one change reverted, we could try this:

- "Changes to **Tagline** were discarded."

ockham committed:
Simplify

paaljoachim commented:
That is a good idea Javier!
Actually having the snackbar list the discarded changes would be very helpful!

paaljoachim commented:
What if the user makes one change and decides not to save it? 

Like so:
![Screenshot 2021-11-12 at 21 20 31](https://user-images.githubusercontent.com/5323259/141529765-ac08aca0-2936-4eca-9eca-61103df33010.jpg)

Should the button with the word "Save" change to a button with the word "Discard"?
As right now the only way to exit the Save panel is to click the X to the right of the Save button.

As there is no way to move onward to the Discard section.

ockham committed:
Filter records that are being saved

ockham commented:
>I only miss the success message indicating that the previous save action was successful. Otherwise, the "What's next" title will be confusing. That title is also important because we aren't showing the snack bar.

Ah right, I meant to add that later but forgot. I'll add it!

> I also think that we could remove the "Continue editing" button. It's too far away from the other actions (the discard button and the close button at the top) and competes with them.

Agree 👍 

> If you all agree to make this change, we should then update the message to something like:
> 
> > **What's next?**
> > Your template still has some unsaved changes.
> > You can select them and discard their changes now, or close the panel and deal with them later.

Will do!

ockham committed:
Remove 'Continue editing' button, adapt wording accordingly

ockham committed:
Add 'Template updated!' message

ockham commented:
I've changed the logic for the discard panel to filter the changes that are currently being saved from the list. (It needs some tweaking, since it always removes site entities, such as title and tagline.)

Furthermore, I've removed the "Continue editing" button, adapted the wording accordingly, and added the "Templated updated!" message. I now remember why I procrastinated on the latter: its bottom padding is too small, and it seems to require quite a bit of effort to get it right 🙄 

![image](https://user-images.githubusercontent.com/96308/141538482-4a4d0d0f-32b9-4ee8-85ef-d826665cd53b.png)

_Still hatin' CSS 😬_ 

I'll continue on Monday.

ockham committed:
Add grey line

ockham committed:
Add saving animation

ockham commented:
> What if the user makes one change and decides not to save it?
> 
> Like so: ![Screenshot 2021-11-12 at 21 20 31](https://user-images.githubusercontent.com/5323259/141529765-ac08aca0-2936-4eca-9eca-61103df33010.jpg)
> 
> Should the button with the word "Save" change to a button with the word "Discard"?

My 2 cents: I don't think that the same button (i.e. same style, same position) should contextually change its function, especially if it's the exact opposite of what it normally does. Users will develop some sort of "muscle memory" and will get used to that button doing one (non-destructive) thing (saving) in most cases; we shouldn't change that to a destructive one in exceptional cases (discarding).

(There might even be some more abstract/general design pattern for this kind of thing, but I'll have to defer to designers on that 😅 )

> As right now the only way to exit the Save panel is to click the X to the right of the Save button.

I think that's sufficient -- it's kind of a "cancel operation" function in all cases.

> As there is no way to move onward to the Discard section.

Right, so the major difference is that we'd be lacking an option that allows us to roll back that one change to the state when it was last saved. I _think_ that's okay, since it's only one change, so the user probably remembers what it was like before, and can revert it manually. (Alternatively, reloading the page should have the same effect.)

Happy to discuss/implement alternatives!

ockham commented:
> * "Changes to **Tagline** were discarded."

This could pose some problems for i18n, but I can add it for now (and we might solicit review from someone more with more i18n experience).

ockham committed:
Add snackbar notification

ockham committed:
Simplify isSaving logic

ockham commented:
> > * "Changes to **Tagline** were discarded."
> 
> This could pose some problems for i18n, but I can add it for now (and we might solicit review from someone more with more i18n experience).

Uh, hold that thought, it turns out to be a bit more involved (not sure if I can use an interpolated component with `createSuccessNotice`). I'll try a simpler version for now ("Change discarded."), since the user is likely aware _which_ change they discarded, since it must be the only checkbox they just ticked...

ockham commented:
> > What if the user makes one change and decides not to save it?
> > Like so: ![Screenshot 2021-11-12 at 21 20 31](https://user-images.githubusercontent.com/5323259/141529765-ac08aca0-2936-4eca-9eca-61103df33010.jpg)
> > Should the button with the word "Save" change to a button with the word "Discard"?
> 
> My 2 cents: [...]
> 
> Happy to discuss/implement alternatives!

Ahh, I had looked at this in the site editor, but I now realize that it's more of a problem in the post editor, where we have the extra step of potentially _publishing_ the post. It's a bit trickier there, since right now, we aren't allowing the user to proceed straight to publishing after saving and discarding. Looks like I'll need to polish this some more...

ockham commented:
Furthermore, the template-centric language ("Template updated", "Your template still has some unsaved changes.") doesn't work in the post editor.

Since we're making these changes to the `@wordpress/editor` package, which is the base for both the post (`@wordpress/edit-post`) and the site editor (`@wordpress/edit-site`) -- and is overall supposed to be editor-type agnostic -- we should change that language to something more generic. E.g. ("Changes saved.", "There are still some unsaved changes.")

ockham committed:
Template-agnostic wording

ockham commented:
Sidenote, the existing e2e test that failed really helped me realize that the current changes don't really work that well in the post editor. We should probably add some e2e test coverage for multi-entity saving in the site editor, too.

ockham commented:
> Ahh, I had looked at this in the site editor, but I now realize that it's more of a problem in the post editor, where we have the extra step of potentially publishing the post. It's a bit trickier there, since right now, we aren't allowing the user to proceed straight to publishing after saving and discarding. Looks like I'll need to polish this some more...

FWIW, here's a screencast illustrating what we're _currently_ (i.e. on `trunk`) doing when there are changes to non-post entities (here: the site title and tagline) in the _post_ editor:

![multi-entity save discard post editor](https://user-images.githubusercontent.com/96308/142066601-dae80f4b-47f5-46ce-ba37-6f7e36d3d80c.gif)

Note that even if we don't save all changes, we're offered to publish (or update, if it wasn't a new post) the post right afterwards. 

How can we reconcile that with the 'Discard changes' panel? 🤔

ockham commented:
Note that this latter scenario also has a bug when clicking "Publish" at the end. I'll file an issue for it tomorrow.

> Uncaught TypeError: t.props.setEntitiesSavedStatesCallback is not a function

ockham commented:
> Note that this latter scenario also has a bug when clicking "Publish" at the end. I'll file an issue for it tomorrow.
> 
> > Uncaught TypeError: t.props.setEntitiesSavedStatesCallback is not a function

Ah, looks like it's already been reported: https://github.com/WordPress/gutenberg/issues/36096

paaljoachim commented:
The Site Editor Discard flow works well!
(When having atleast one item that is to be saved.)

https://user-images.githubusercontent.com/5323259/142516266-e2b73ca6-7ccc-4a28-b422-dd056816cd00.mp4

---------
----
-----


**Moving on to....**

Making one change.

### Site Editor

Step 1.
This is what it looks like in the Site Editor:
![Screenshot 2021-11-19 at 01 07 44](https://user-images.githubusercontent.com/5323259/142516311-78851892-fc61-4863-b1c3-b0d84c106c6d.jpg)

Step 2.
Clicking Save and unchecking Title.
![Screenshot 2021-11-19 at 01 10 25](https://user-images.githubusercontent.com/5323259/142516482-179a2f2f-dcec-42fa-9224-c7b0bb4f650f.jpg)

We get trapped. No way to discard the change to the Title.

-

### Post Editor

Step 1.
This is what it looks like in the Post Editor:
![Screenshot 2021-11-19 at 01 09 32](https://user-images.githubusercontent.com/5323259/142516367-07a8f927-e313-4a2d-96a6-8e3a0716c327.jpg)

Step 2.
Clicking Update and unchecking the Reusable block.
![Screenshot 2021-11-19 at 01 13 15](https://user-images.githubusercontent.com/5323259/142516782-7607e15a-f0dc-4e43-8110-7432682c665d.jpg)

We get trapped. No way to discard the change to the Reusable block.

----

(First of I would suggest to remove the dot seen in the Update button (Post Editor). It would be nice to keep the Post Editor flow as similar to the Site Editor save flow as possible.)

We need a discard flow for when no item is to be saved. How would we move on to the discard panel when no item has been checked and the Save button is greyed out? 

Here is one suggestion. Unchecking so no item it so be saved the Discard button could show up below. 
It can be a bit confusing. As do one need to check both items and then click the Discard button, or can one just click the discard button to discard and revert both items?

![No-save-Discard-Changes](https://user-images.githubusercontent.com/5323259/142518377-303b8fd8-132e-4a13-bfe1-b846592a9ac7.jpg)


@javierarce Javier what would you suggest?

paaljoachim commented:
I see this PR has stopped up. It would be great to get some movement again here!
Thanks!

javierarce commented:
👋 Chiming in just to say that I'll try to review this PR this week.

ockham commented:
> I see this PR has stopped up. It would be great to get some movement again here!
> Thanks!

Yeah, sorry about that 😕  As stated on the [related issue](https://github.com/WordPress/gutenberg/issues/36096), the multi-entity saving code is pretty complex, and I had to prioritize some other WP 5.9 stuff.

As for this PR, it's even worse: There's the bug (#36096), there's the UX questions (such as [this](https://github.com/WordPress/gutenberg/pull/36185#issuecomment-967485096)), and then, there's different contexts (the premises for multi-entity saving in the site editor are quite different from the post editor). We might actually need more UX input, especially for the post editor context. 

> 👋 Chiming in just to say that I'll try to review this PR this week.

@javierarce Thanks a lot for offering -- I'm just not quite sure this is ready for prime-time yet. Maybe do try it out both in the site and post editors, and have a look at @paaljoachim's comments above; we might need some more UX guidance 😅

paaljoachim commented:
Hey Bernie @ockham 

I am checking in. Can we get a status update as to the next steps needed?
Thanks!

Mamaduka commented:
Doing some old PR cleanup. I'm going to close this as "stale".
